A36-xds-for-servers.md

```diff
@@ -0,0 +1,316 @@
+inform the application of initial communication failures. If the user does not
+configure the API, gRPC should log the errors at a default-visible log level. If
+the xds bootstrap is missing or invalid, implementations would ideally fail
+server startup, but a permanent hang is also acceptable. "Initial configuration"
+does not have to include server credentials.
+
+The `GRPC_XDS_BOOTSTRAP` file will be enhanced to have a new field:
+```json
+{
+  "grpc_server_resource_name_id": "example/resource",
+  // ...
+}
+```
+
+XdsServer will use the normal `XdsClient` to communicate with the xDS server.
+The XdsServer will pass the listening address (commonly using a wildcard IP
+address, like `::` or `0.0.0.0`) to a new XdsClient API to start a watch for a
+`envoy.config.listener.v3.Listener` resource. XdsClient will concatenate the
+value of `grpc_server_resource_name_id` from the bootstrap,
+`?udpa.resource.listening_address=`, and the listening address to construct the
+resource name. For example, with an address of `[::]:80` and the example
+bootstrap field above, the resource name would be
+`example/resource?udpa.resource.listening_address=[::]:80`. Note that the `[]`
+is _not_ percent encoded, as this string is not a URI.
+
+The xDS-returned Listener must have an [`address`][Listener.address] that
+matches the listening address provided. The Listener's `address` would be a TCP
+`SocketAddress` with matching `address` and `port_value`. The XdsClient must
+NACK the resource if the address does not match.
+
+Although `Filter`s will not be observed initially, the `FilterChain` contains
+data that may be used. When looking for a FilterChain, the standard matching
+logic must be used. Each `filter_chain_match` of the repeated
+[`filter_chains`][Listener.filter_chains] should be checked and if none match,
+the `default_filter_chain` must be used.  The following is a snippet of [all
+current fields of FilterChainMatch][FilterChainMatch], and if they will be
+handled specially:
+
+[Listener.address]: https://github.com/envoyproxy/envoy/blob/11dee4e9c37c244f9de4bc339fdc8695e5de2c5d/api/envoy/config/listener/v3/listener.proto#L108
+[Listener.filter_chains]: https://github.com/envoyproxy/envoy/blob/11dee4e9c37c244f9de4bc339fdc8695e5de2c5d/api/envoy/config/listener/v3/listener.proto#L117
+[FilterChainMatch]: https://github.com/envoyproxy/envoy/blob/928a62b7a12c4d87ce215a7c4ebd376f69c2e080/api/envoy/config/listener/v3/listener_components.proto#L85
+
+```
+message FilterChainMatch {
+  enum ConnectionSourceType {
+    ANY = 0;
+    SAME_IP_OR_LOOPBACK = 1;
+    EXTERNAL = 2;
+  }
+  google.protobuf.UInt32Value destination_port = 8; // Always fail match
+  repeated core.v3.CidrRange prefix_ranges = 3;
+  ConnectionSourceType source_type = 12;
+  repeated core.v3.CidrRange source_prefix_ranges = 6;
+  repeated uint32 source_ports = 7;
+  repeated string server_names = 11; // May always fail match if TLS unsupported
```

Comment:
Doh. This appears to rely on tls_inspector. https://www.envoyproxy.io/docs/envoy/latest/faq/configuration/sni . The proto documentation wasn't quite as clear: "Those values will be compared against the server names of a new connection, when detected by one of the listener filters."

transport_protocol is only useful for sniffing, which we aren't interested in supporting. application_protocols could potentially be useful in the future, but not so much now (always h2). But server_names is more useful... Need to decide if/how we handle it.

Reply:
I vote against supporting this since we don't have tls_inspector kind of functionality. It's not worth the trouble to add this support esp considering there might not be use-cases.